src/posix_dbus.c

```diff
@@ -0,0 +1,326 @@
+
+    //Check for errors
+    if(dbus_error_is_set(&_glfw.dbus.error) || !_glfw.dbus.connection)
+    {
+        if(dbus_error_is_set(&_glfw.dbus.error))
+            dbus_error_free(&_glfw.dbus.error);
+
+        _glfwInputError(GLFW_PLATFORM_ERROR, "Failed to connect to DBus");
+
+        dbus_connection_unref(_glfw.dbus.connection);
+        _glfw.dbus.connection = NULL;
+    }
+    else
+    {
+        //Request name
+        const int res = dbus_bus_request_name(_glfw.dbus.connection, "org.glfw", DBUS_NAME_FLAG_REPLACE_EXISTING, &_glfw.dbus.error);
+
+        //Check for errors
+        if(dbus_error_is_set(&_glfw.dbus.error) || res != DBUS_REQUEST_NAME_REPLY_PRIMARY_OWNER)
+        {
+            if(dbus_error_is_set(&_glfw.dbus.error))
+                dbus_error_free(&_glfw.dbus.error);
+
+            _glfwInputError(GLFW_PLATFORM_ERROR, "Failed to request DBus name");
+
+                dbus_connection_unref(_glfw.dbus.connection);
+                _glfw.dbus.connection = NULL;
+        }
+    }
+}
+
+void _glfwTerminateDBusPOSIX(void)
+{
+    if (_glfw.dbus.connection)
+    {
+        dbus_connection_unref(_glfw.dbus.connection);
+        _glfw.dbus.connection = NULL;
+    }
+
+    if (_glfw.dbus.handle)
+    {
+        _glfwPlatformFreeModule(_glfw.dbus.handle);
+        _glfw.dbus.handle = NULL;
+    }
+}
+
+void _glfwUpdateTaskbarProgressDBusPOSIX(dbus_bool_t progressVisible, double progressValue)
+{
+    if(!_glfw.dbus.handle || !_glfw.dbus.connection)
+    {
+        _glfwInputError(GLFW_FEATURE_UNAVAILABLE, "POSIX: No DBus connection open to set taskbar progress");
+        return;
+    }
+
+    //Signal signature:
+    //signal com.canonical.Unity.LauncherEntry.Update (in s app_uri, in a{sv} properties)
+
+    struct DBusMessageIter args;
+    memset(&args, 0, sizeof(args));
+
+    //Get name of the running executable
+    char exeName[PATH_MAX];
+    memset(exeName, 0, sizeof(char) * PATH_MAX);
+    if(readlink("/proc/self/exe", exeName, PATH_MAX) == -1)
+    {
+        _glfwInputError(GLFW_PLATFORM_ERROR, "Failed to get name of the running executable");
+        return;
+    }
+    char* exeNameEnd = strchr(exeName, '\0');
+    char* lastFound = strrchr(exeName, '/');
+    if(!lastFound || !exeNameEnd)
+    {
+        _glfwInputError(GLFW_PLATFORM_ERROR, "Failed to get name of the running executable");
+        return;
+    }
+    unsigned int exeNameLength = (exeNameEnd - lastFound) - 1;
+
+    //Create our final desktop file uri
+    unsigned int desktopFileLength = strlen("application://") + exeNameLength + strlen(".desktop") + 1;
+    char desktopFile[desktopFileLength];
+    memset(desktopFile, 0, sizeof(char) * desktopFileLength);
+    strcpy(desktopFile, "application://");
+    memcpy(desktopFile + strlen("application://"), lastFound + 1, exeNameLength);
+    strcpy(desktopFile + strlen("application://") + (exeNameLength), ".desktop");
+    desktopFile[desktopFileLength - 1] = '\0';
```

Comment:
Is it necessary to build the `desktopFile` on every invocation? Can this be cached in `_glfw` (lazy loading)?

Reply:
We can lazy load it.
Should we load it during `_glfwInitDBusPOSIX()` or check on each invocation of `_glfwUpdateTaskbarProgressDBusPOSIX()` and `_glfwUpdateBadgeDBusPOSIX()` if its already loaded?

Reply:
`_glfwInitDBusPOSIX()` sounds fine.